### PR TITLE
add a Size field to Query's Result

### PR DESCRIPTION
**Pull request by MichaelMure**
Rational is that some datastore can easilty get the size of a value when listing only keys, so we might as well have a way to pass that. One way this could be useful is when implementing a cache for Has/GetSize.

### CODE REVIEW

File: query/query.go
Comment (Stebalien) [inline query/query.go]: So, a value can _also_ just be empty. We'd have to use -1 to mean "not set" (or a `HasSize` flag).

What if we:

1. Added a field to the query to _request_ the size (like we do with expirations).
2. Always included the size if requested, potentially doing extra work.

This is a breaking change for datastore implementations but it's a pretty minor breaking change.

**Comment by Stebalien** [comment]
Hm. I just wanted this when messing with `ipfs repo stat` (I wanted to be able to report sizes for various record types).

File: query/query.go
Comment (MichaelMure) [reply query/query.go]: That works for me. A datastore layer can also hijack the query and enable the flag if it's needed (say, a caching layer) or considered free (say, S3).

**Comment by MichaelMure** [comment]
I changed the semantic so that:
- the calling side can opt-in to always have the size set, even with `KeysOnly`, even it that implies a performance cost
- a datastore implementation can optionally always return the size if it doesn't imply a performance cost

I feel like that the best compromise. What do you think ?

**Review by Stebalien — APPROVED** [review]
LGTM!

**Comment by Stebalien** [comment]
Could you file pull requests against the other datastores before we merge this?

* [x] go-ds-badger
* [ ] ~go-ds-bolt~ broken
* [x] go-ds-flatfs
* [x] go-ds-leveldb
* [x] go-ds-redis
* [x] go-ds-s3

**Review by Stebalien — CHANGES_REQUESTED** [review]
Actually, this needs to update the mount datastore. We don't need to forward the `RequestSize` to child queries.

While we're at it, we should add tests to `test/suite.go`.

**Comment by MichaelMure** [comment]
> Actually, this needs to update the mount datastore. We don't need to forward the RequestSize to child queries.

Not sure I follow, It doesn't at the moment, but it should, right ?

**Comment by Stebalien** [comment]
Ah, sorry, yes, we _need_ to forward ReturnsSize.

Also, I noticed the variable is currently `ReturnsSize`, it should be `ReturnSizes` (we're _asking_ the datastore to return the sizes).

**Comment by MichaelMure** [comment]
- fixed the mount datastore
- renamed ReturnsSize to ReturnsSizes
- added a test for ReturnsSizes

Should be good to go now

**Comment by MichaelMure** [comment]
> Could you file pull requests against the other datastores before we merge this?
> 
* [X]  go-ds-badger : issue opened, https://github.com/ipfs/go-ds-badger/pull/76
* [ ]  go-ds-bolt : can't file issue there, first step: https://github.com/ipfs/go-ds-bolt/pull/4
* [X]  go-ds-flatfs : issue opened, https://github.com/ipfs/go-ds-flatfs/pull/59
* [X]  go-ds-leveldb : issue opened, https://github.com/ipfs/go-ds-leveldb/pull/35
* [X]  go-ds-redis : issue opened, but also **doesn't support query to begin with**
* [X]  go-ds-s3: https://github.com/ipfs/go-ds-s3/pull/45

**Comment by Stebalien** [comment]
Unfortunately, I can't merge this until we actually have PRs for those issues. Otherwise, all those datastores will return invalid results.

**Comment by MichaelMure** [comment]
Any chance the respective maintainers could help here ?

**Comment by Stebalien** [comment]
Unfortunately, I am the de facto maintainer.

**Comment by MichaelMure** [comment]
I did a pass on all of them but `go-ds-bolt` is problematic as it's broken at the moment.

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
All merged and updated.

**Pull request merged**
